homeassistant/components/sms/gateway.py

```diff
@@ -15,6 +18,120 @@ class Gateway:
     def __init__(self, worker, hass):
         """Initialize the sms gateway."""
         self._worker = worker
+        self._loop = asyncio.get_running_loop()
+        self._hass = hass
+
+    async def init_async(self):
+        """Initialize the sms gateway asyncronisly."""
+        try:
+            await self._worker.SetIncomingSMSAsync()
+        except gammu.ERR_NOTSUPPORTED:
+            _LOGGER.warning("Your phone does not support incoming SMS notifications!")
+        else:
+            await self._worker.SetIncomingCallbackAsync(self.sms_callback)
+
+    def sms_callback(self, state_machine, callback_type, callback_data):
+        """Receive notification about incoming event.
+
+        @param state_machine: state machine which invoked action
+        @type state_machine: gammu.StateMachine
+        @param callback_type: type of action, one of Call, SMS, CB, USSD
+        @type callback_type: string
+        @param data: event data
+        @type data: hash
+        """
+        _LOGGER.debug(
+            f"Received incoming event type:{callback_type}, data:{callback_data}"
+        )
+        entries = self.get_and_delete_all_sms(state_machine)
+        _LOGGER.debug(f"SMS entries:{entries}")
+        data = list()
+
+        for entry in entries:
+            v = gammu.DecodeSMS(entry)
+            message = entry[0]
+            _LOGGER.debug(f"Processing sms {message}, decoded: {v}")
+            if v is None:
+                text = message["Text"]
+            else:
+                text = ""
+                for e in v["Entries"]:
+                    if e["Buffer"] is not None:
+                        text = text + e["Buffer"]
+
+            event_data = dict(
+                phone=message["Number"], date=str(message["DateTime"]), message=text
+            )
+
+            _LOGGER.debug(f"Append event data:{event_data}")
+            data.append(event_data)
+
+        self._loop.call_soon_threadsafe(self._notify_incoming_sms, data)
+
+    def get_and_delete_all_sms(self, state_machine, force=False):
+        """Read and delete all SMS in the modem."""
+        # Read SMS memory status ...
+        memory = state_machine.GetSMSStatus()
```

Comment:
Is this doing I/O ?

Reply:
> Is this doing I/O ?

Yes, the gammu worker thread is doing I/O directly on the serial device.